README.md
## Hyperium Addon Workspace

Hyperium uses a system called addons, which allow you to add whatever you'd like to the client.

### Getting Started

To get started, first you want to either clone or download the addon workspace from here,
open it in an IDE that supports Java or Kotlin (preferably [IntelliJ](https://www.jetbrains.com/idea/)), and import the `build.gradle` file.

Likely, at first you won't have access to the decompiled Minecraft code. To fix this, all you have to do
is open the Gradle tab on the right side of your screen. Open the dropdown menu, `Tasks`, `forgegradle`, and run `setupDecompWorkspace`.

After this, refresh your gradle deps using the button that looks like two arrows spinning around eachother. 
Once you've done that, you can open `Tasks`, `forgegradle`, and run `genIntellijRuns`. This is going to create the base for
running the client once you're ready to test something.

Test to see if it setup properly by going into the top right and seeing a dropdown menu saying `Minecraft Client`, beside that is a green arrow.
Press the arrow, if it starts up, you're all good, otherwise, click on the dropdown menu, and where it says `Use classpath of module: %WORKSPACE_NAME%`,
you will want to change that to be `%WORKSPACE_NAME%.main`. An example of this would be `ExampleAddon.main`.

At this point, you should be good to go.

If you wish to login to your account to test something on a server, open the dropdown menu saying Minecraft Client, and insert
 `--username (username/email) --password (password)` then run the client.

### Addon Manifests

All of your addon information will go into `resources/addon.json`.

Listed below is all of the manifest information, and an example provided.

* **name**: The name of your addon. **ExampleAddon**
* **version**: The version of your addon. **1.0**
* **mainClass**: The path to the main class of your addon. **AddonExample** 
* **desc**: A description of your addon. **This is an addon that is made for an example.**
* **versionCode**: MUST match up to the client's versionCode. **The current versionCode is available in `AddonMinecraftBootstrap#VERSION_CODE`**


### Examples

An example of a completed addon.json file is below.

```json
{
  "name":"Addon Example",
  "version":"1.0",
  "mainClass":"AddonExample",
  "desc":"Hyperium Addon Example",
  "versionCode": "1.0"
}
```

An example of a basic addon is below.

```java
// All addons must implement IAddon
public class AddonExample implements IAddon {

    /**
     * Invoked once the plugin has successfully loaded
     * {@link cc.hyperium.internal.addons.AddonMinecraftBootstrap#init}
     */
    @Override
    public void onLoad() {
        // Log that the addon has loaded
        Hyperium.LOGGER.info("Successfully loaded Addon!");
        
        // Register the class to the EventBus
        EventBus.INSTANCE.register(this);
    }
    
    @InvokeEvent
    public void onInitialization(InitializationEvent event) {
        // Register a command on initialization
        Hyperium.INSTANCE.getHandlers().getCommandHandler().registerCommand(new AddonCommand());
    }
    
    @InvokeEvent
    public void onChatMessage(ChatEvent event) {
        // If any chat message contains "secret message!"
        if (event.getChat().getUnformattedText().contains("secret message!")) {
            //Toggle fullscreen
            Minecraft.getMinecraft().toggleFullscreen();
        }
    }
    
    /**
     * Invoked once the game has been closed
     * this is executed at the start of {@link net.minecraft.client.Minecraft#shutdown}
     */
    @Override
    public void onClose() {
        // Log that the addon is being closed
        Hyperium.LOGGER.info("Closing...");
    }

    /**
     * Invoked on debug call. Can be used to add things into crash reports
     * <p>
     * This does not need to be overriden if it's not needed
     */
    @Override
    public void sendDebugInfo() {
    }
}
```

An example of a basic addon command is below.

```java
public class AddonCommand implements BaseCommand {
    
    /**
     * Gets the name of the command
     */
    @Override
    public String getName() {
        return "addon_example";
    }

    /**
     * Gets the usage string for the command.
     */
    @Override
    public String getUsage() {
        return "/" + getName();
    }

    /**
     * Callback when the command is invoked
     *
     * @param args
     * @throws CommandException for errors inside the command, these errors
     *                          will log directly to the players chat (without a prefix)
     */
    @Override
    public void onExecute(String[] args) throws CommandException {
        Minecraft.getMinecraft().toggleFullscreen();
    }
}
```

CHANGELOG.md
In order to comply with the license, here are the changes that this fork has made to the codebase:

* Fix bugs with command handler
* Use hyperiumjailbreak snapshots
* Clean up build.gradle
* A bit of code cleanup
* Ensure settings are saved


src/main/java/club/chachy/oofmod/OofMod.java
package club.chachy.oofmod;

import cc.hyperium.Hyperium;
import cc.hyperium.event.EventBus;
import cc.hyperium.event.InvokeEvent;
import cc.hyperium.event.client.InitializationEvent;
import cc.hyperium.internal.addons.IAddon;
import club.chachy.oofmod.commands.OpenMenuCommand;
import club.chachy.oofmod.listeners.OofModListener;
import club.chachy.oofmod.settings.Settings;

public class OofMod implements IAddon {
    public static OofMod INSTANCE = new OofMod();
    private Settings settings = new Settings();

    /**
     * Invoked once the plugin has successfully loaded
     * {@link cc.hyperium.internal.addons.AddonMinecraftBootstrap#init}
     */
    @Override
    public void onLoad() {
        Hyperium.LOGGER.info("Successfully loaded OofAddonV2!");
        EventBus.INSTANCE.register(this);
    }

    @InvokeEvent
    public void onInitialization(InitializationEvent event) {
        EventBus.INSTANCE.register(this);
        Hyperium.CONFIG.register(Settings.INSTANCE);
        EventBus.INSTANCE.register(new OofModListener());
        Hyperium.INSTANCE.getHandlers().getCommandHandler().registerCommand(new OpenMenuCommand());
    }

    /**
     * Invoked once the game has been closed
     * this is executed at the start of {@link net.minecraft.client.Minecraft#shutdown}
     */
    @Override
    public void onClose() {
        // Log that the addon is being closed
        Hyperium.LOGGER.info("Closing OofAddonV2");
    }

    public Settings getSettings() {
        return settings;
    }
}


src/main/java/club/chachy/oofmod/settings/Settings.java
package club.chachy.oofmod.settings;

import cc.hyperium.config.ConfigOpt;

import java.io.DataInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;

public class Settings {
    public static Settings INSTANCE = new Settings();
    private File soundsFolder = new File("config/oofmod/sounds");
    @ConfigOpt
    private String selectedSoundName = "oof";
    @ConfigOpt
    private File selectedSound = new File(soundsFolder.getPath() + "/" + selectedSoundName + ".wav");
    @ConfigOpt
    private float volume = 1.0f;
    @ConfigOpt
    private boolean enabled = true;

    public Settings() {
        if (!soundsFolder.exists()) {
            soundsFolder.mkdirs();
        }
        downloadDefaultSound();
    }

    public ArrayList<File> getSounds() {
        ArrayList<File> result = new ArrayList<>();
        for (File file : soundsFolder.listFiles()) {
            if (file.getName().endsWith(".wav")) {
                result.add(file);
            }
        }
        return result;
    }

    public void setSelectedSoundName(String soundName) {
        selectedSoundName = soundName;
        selectedSound = new File(soundsFolder.getPath() + "/" + selectedSoundName);
    }


    public File getSelectedSound() {
        return this.selectedSound;
    }

    public boolean isEnabled() {
        return this.enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public float getVolume() {
        return this.volume;
    }

    public void setVolume(float volume) {
        this.volume = volume;
    }

    public String getSelectedSoundName() {
        return selectedSoundName;
    }

    private void downloadDefaultSound() {
        File defaultSound = new File("config/oofmod/sounds/oof.wav");
        if (defaultSound.exists()) {
            return;
        }
        URL url;
        URLConnection con;
        DataInputStream dis;
        FileOutputStream fos;
        byte[] fileData;
        try {
            url = new URL("https://oofmodsound.powns.dev/oof.wav"); //File Location goes here
            con = url.openConnection(); // open the url connection.
            dis = new DataInputStream(con.getInputStream());
            fileData = new byte[con.getContentLength()];
            for (int q = 0; q < fileData.length; q++) {
                fileData[q] = dis.readByte();
            }
            dis.close(); // close the data input stream
            fos = new FileOutputStream(new File(soundsFolder.getPath() + "/oof.wav")); //FILE Save Location goes here
            fos.write(fileData);  // write out the file we want to save.
            fos.close(); // close the output stream writer
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}


src/main/java/club/chachy/oofmod/commands/OpenMenuCommand.java
package club.chachy.oofmod.commands;

import cc.hyperium.Hyperium;
import cc.hyperium.commands.BaseCommand;
import club.chachy.oofmod.OofMod;
import club.chachy.oofmod.gui.OofGui;
import net.minecraft.client.Minecraft;
import net.minecraft.event.ClickEvent;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.EnumChatFormatting;
import net.minecraft.util.IChatComponent;

public class OpenMenuCommand implements BaseCommand {

    /**
     * Gets the name of the command
     */
    @Override
    public String getName() {
        return "oofmod";
    }

    /**
     * Gets the usage string for the command.
     */
    @Override
    public String getUsage() {
        return "/" + getName();
    }

    /**
     * Callback when the command is invoked
     *
     * @param args
     */
    @Override
    public void onExecute(String[] args) {
        if (OofMod.INSTANCE.getSettings().getSounds().size() == 0) {
            Minecraft.getMinecraft().thePlayer.addChatMessage(new ChatComponentText(EnumChatFormatting.DARK_RED + "------[OOFMOD ERROR]------"));
            Minecraft.getMinecraft().thePlayer.addChatMessage(new ChatComponentText(EnumChatFormatting.GOLD + ".minecraft/config/oofmod/sounds is empty!"));
            Minecraft.getMinecraft().thePlayer.addChatMessage(new ChatComponentText(EnumChatFormatting.GOLD + "Add some files before using this command."));

            IChatComponent download = new ChatComponentText(EnumChatFormatting.GOLD + "" + EnumChatFormatting.UNDERLINE + "CLICK HERE to download oof.wav.");
            download.getChatStyle().setChatClickEvent(new ClickEvent(ClickEvent.Action.OPEN_URL, "https://www.mediafire.com/file/l7tv8u9dsleq7ul/oof.wav/file"));
            Minecraft.getMinecraft().thePlayer.addChatMessage(download);
            Minecraft.getMinecraft().thePlayer.addChatMessage(new ChatComponentText(EnumChatFormatting.DARK_RED + "-------------------------"));
            return;
        }
        Hyperium.INSTANCE.getHandlers().getGuiDisplayHandler().setDisplayNextTick(new OofGui());
    }
}


src/main/java/club/chachy/oofmod/listeners/OofModListener.java
package club.chachy.oofmod.listeners;

import cc.hyperium.event.InvokeEvent;
import cc.hyperium.event.network.chat.ServerChatEvent;
import club.chachy.oofmod.OofMod;
import com.mojang.authlib.GameProfile;
import me.semx11.autotip.event.impl.EventClientTick;
import net.minecraft.client.Minecraft;
import net.minecraft.client.entity.EntityPlayerSP;
import net.minecraft.client.network.NetHandlerPlayClient;
import net.minecraft.client.network.NetworkPlayerInfo;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;
import javax.sound.sampled.FloatControl;
import java.io.File;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class OofModListener {
    private Minecraft mc = Minecraft.getMinecraft();

    private String nameToCheck = "";

    private OofMod mod = OofMod.INSTANCE;

    @InvokeEvent
    public void onDeathMessage(ServerChatEvent event) throws Exception {
        if (nameToCheck.isEmpty()) {
            nameToCheck = mc.thePlayer.getName();
        }
        String line = event.getChat().getUnformattedText();
        if (!mod.getSettings().isEnabled() || line.split(" ").length == 0) {
            return;
        }
        String killMessageRegex = "(\\w{1,16}).+ (by|of|to|for|with) (" + nameToCheck + ")";
        String usernamePatternRegex = "^[a-zA-Z0-9_-]{3,16}$";
        Pattern killMessagePattern = Pattern.compile(killMessageRegex);
        Pattern usernamePattern = Pattern.compile(usernamePatternRegex);
        Matcher killMessageMatcher = killMessagePattern.matcher(line);
        Matcher usernameMatcher = usernamePattern.matcher(line.split(" ")[0]);
        if (usernameMatcher.matches() && killMessageMatcher.find()) {
            String killed = killMessageMatcher.group(1);
            if (!killed.equals(nameToCheck)) {
                playOofSound();
            }
        }
    }

    @InvokeEvent
    public void profileCheck(EventClientTick event) {
        EntityPlayerSP player = Minecraft.getMinecraft().thePlayer;
        if (player == null) {
            return;
        }
        NetHandlerPlayClient sendQueue = player.sendQueue;
        if (sendQueue == null) {
            return;
        }
        for (NetworkPlayerInfo networkPlayerInfo : sendQueue.getPlayerInfoMap()) {
            GameProfile gameProfile = networkPlayerInfo.getGameProfile();
            if (gameProfile.getId() != null && gameProfile.getId().equals(Minecraft.getMinecraft().getSession().getProfile().getId())) {
                nameToCheck = gameProfile.getName();
                break;
            }
        }
    }

    private void playOofSound() throws Exception {
        if (!mod.getSettings().getSelectedSound().exists()) {
            return;
        }
        File soundFile = mod.getSettings().getSelectedSound();
        AudioInputStream audioIn = AudioSystem.getAudioInputStream(soundFile.toURI().toURL());
        Clip clip = AudioSystem.getClip();
        clip.open(audioIn);
        FloatControl gainControl = (FloatControl) clip.getControl(FloatControl.Type.MASTER_GAIN);
        gainControl.setValue(mod.getSettings().getVolume() - 30.0f);
        clip.start();
    }
}



src/main/java/club/chachy/oofmod/gui/helper/GuiTransButton.java
package club.chachy.oofmod.gui.helper;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.GuiButton;
import net.minecraft.client.renderer.OpenGlHelper;
import org.lwjgl.opengl.GL11;

public class GuiTransButton extends GuiButton {
    private int packedFGColour;

    public GuiTransButton(int id, int x, int y, int width, int height, String displayString) {
        super(id, x, y, width, height, displayString);
    }

    public void drawButton(Minecraft mc, int mouseX, int mouseY) {
        if (visible) {
            FontRenderer fontrenderer = mc.fontRendererObj;
            mc.getTextureManager().bindTexture(GuiTransButton.buttonTextures);
            GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
            hovered = (mouseX >= xPosition && mouseY >= yPosition && mouseX < xPosition + width && mouseY < yPosition + height);
            int k = getHoverState(hovered);
            GL11.glEnable(3042);
            OpenGlHelper.glBlendFunc(770, 771, 1, 0);
            GL11.glBlendFunc(770, 771);
            drawRect(xPosition + 1, yPosition + 1, xPosition + width - 1, yPosition + height - 1, getButtonColor());
            mouseDragged(mc, mouseX, mouseY);
            int l = 14737632;
            if (this.packedFGColour != 0) {
                l = packedFGColour;
            } else if (!enabled) {
                l = 10526880;
            } else if (hovered) {
                l = 16777120;
            }
            drawCenteredString(fontrenderer, displayString, xPosition + width / 2, yPosition + (height - 8) / 2, l);
        }
    }

    private int getButtonColor() {
        switch (super.getHoverState(super.hovered)) {
            case 0: {
                return getColorWithAlpha(8168374, 50);
            }
            case 1: {
                return getColorWithAlpha(8168374, 100);
            }
            case 2: {
                return getColorWithAlpha(8168374, 150);
            }
            default: {
                return 0;
            }
        }
    }

    private int getColorWithAlpha(int rgb, int a) {
        int r = rgb >> 16 & 0xFF;
        int g = rgb >> 8 & 0xFF;
        int b = rgb & 0xFF;
        return a << 24 | r << 16 | g << 8 | b;
    }
}



src/main/java/club/chachy/oofmod/gui/helper/GuiSlideControl.java
package club.chachy.oofmod.gui.helper;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.GuiButton;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.OpenGlHelper;
import org.lwjgl.opengl.GL11;

import java.text.DecimalFormat;

public class GuiSlideControl extends GuiButton {
    private static DecimalFormat numFormat = new DecimalFormat("#.00");
    public String label;
    public float curValue;
    public float minValue;
    public float maxValue;
    public boolean isSliding;
    public boolean useIntegers;
    public int packedFGColour;


    public GuiSlideControl(int id, int x, int y, int width, int height, String displayString, float minVal, float maxVal, float curVal, boolean useInts) {
        super(id, x, y, width, height, useInts ? (displayString + (int) curVal) : (displayString + GuiSlideControl.numFormat.format(curVal)));
        label = displayString;
        minValue = minVal;
        maxValue = maxVal;
        curValue = (curVal - minVal) / (maxVal - minVal);
        useIntegers = useInts;
    }

    public float getValueAsFloat() {
        return (maxValue - minValue) * curValue + minValue;
    }

    public int getValueAsInt() {
        return (int) ((maxValue - minValue) * curValue + minValue);
    }

    protected float roundValue(float value) {
        value = 0.01f * Math.round(value / 0.01f);
        return value;
    }

    public String getLabel() {
        if (useIntegers) {
            return label + getValueAsInt();
        }
        return label + GuiSlideControl.numFormat.format(getValueAsFloat());
    }

    protected void setLabel() {
        displayString = getLabel();
    }

    public int getHoverState(boolean isMouseOver) {
        return 0;
    }

    protected void mouseDragged(Minecraft mc, int mousePosX, int mousePosY) {
        if (visible) {
            if (isSliding) {
                curValue = (mousePosX - (xPosition + 4.0f)) / (width - 8.0f);
                if (curValue < 0.0f) {
                    curValue = 0.0f;
                }
                if (curValue > 1.0f) {
                    curValue = 1.0f;
                }
                setLabel();
            }
            GlStateManager.color(1.0f, 1.0f, 1.0f, 1.0f);
            drawRect(xPosition + 1, yPosition + 1, xPosition + (int) (curValue * (width - 8)) + 7, yPosition + height - 1, getColorWithAlpha(8168374, 100));
            drawRect(xPosition + (int) (curValue * (width - 8)) + 1, yPosition + 1, xPosition + (int) (curValue * (width - 8)) + 7, yPosition + height - 1, getColorWithAlpha(8168374, 200));
        }
    }

    public boolean mousePressed(Minecraft mc, int mousePosX, int mousePosY) {
        if (!super.mousePressed(mc, mousePosX, mousePosY)) {
            return false;
        }
        curValue = (mousePosX - (xPosition + 4.0f)) / (width - 8.0f);
        if (curValue < 0.0f) {
            curValue = 0.0f;
        }
        if (curValue > 1.0f) {
            curValue = 1.0f;
        }
        setLabel();
        if (isSliding) {
            return isSliding = false;
        }
        return isSliding = true;
    }

    public void mouseReleased(int mousePosX, int mousePosY) {
        isSliding = false;
    }

    public void drawButton(Minecraft mc, int mouseX, int mouseY) {
        if (visible) {
            FontRenderer fontrenderer = mc.fontRendererObj;
            mc.getTextureManager().bindTexture(GuiSlideControl.buttonTextures);
            GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
            hovered = (mouseX >= xPosition && mouseY >= yPosition && mouseX < xPosition + width && mouseY < yPosition + height);
            int k = getHoverState(visible);
            GL11.glEnable(3042);
            OpenGlHelper.glBlendFunc(770, 771, 1, 0);
            GL11.glBlendFunc(770, 771);
            drawRect(xPosition + 1, yPosition + 1, xPosition + width - 1, yPosition + height - 1, getColorWithAlpha(8168374, 50));
            mouseDragged(mc, mouseX, mouseY);
            int l = 14737632;
            if (packedFGColour != 0) {
                l = packedFGColour;
            } else if (!enabled) {
                l = 10526880;
            } else if (hovered) {
                l = 16777120;
            }
            drawCenteredString(fontrenderer, displayString, xPosition + width / 2, yPosition + (height - 8) / 2, l);
            GL11.glColor3f(1.0f, 1.0f, 1.0f);
        }
    }

    private int getColorWithAlpha(int rgb, int a) {
        int r = rgb >> 16 & 0xFF;
        int g = rgb >> 8 & 0xFF;
        int b = rgb & 0xFF;
        return a << 24 | r << 16 | g << 8 | b;
    }
}



src/main/java/club/chachy/oofmod/gui/OofGui.java
package club.chachy.oofmod.gui;

import cc.hyperium.Hyperium;
import club.chachy.oofmod.OofMod;
import club.chachy.oofmod.gui.helper.GuiSlideControl;
import club.chachy.oofmod.gui.helper.GuiTransButton;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.gui.GuiButton;
import net.minecraft.client.gui.GuiScreen;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.WorldRenderer;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.util.EnumChatFormatting;
import net.minecraft.util.ResourceLocation;
import org.lwjgl.opengl.GL11;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;
import javax.sound.sampled.FloatControl;
import java.awt.*;
import java.io.File;
import java.util.ArrayList;

public class OofGui extends GuiScreen {
    private OofMod mod = OofMod.INSTANCE;
    private int currentIndex = mod.getSettings().getSounds().indexOf(mod.getSettings().getSelectedSound());
    private GuiSlideControl volume;


    public void initGui() {
        buttonList.add(new GuiTransButton(0, getCenter() - 46, getRowPos(2) + 10, 16, 16, "<"));
        buttonList.add(new GuiTransButton(1, getCenter() + 30, getRowPos(2) + 10, 16, 16, ">"));
        buttonList.add(new GuiTransButton(2, getCenter() - 23, getRowPos(1) + 8, 46, 55, ""));
        buttonList.add(new GuiTransButton(3, getCenter() - 45, getRowPos(4) + 10, 90, 18, mod.getSettings().isEnabled() ? (EnumChatFormatting.GREEN + "Enabled") : (EnumChatFormatting.RED + "Disabled")));
        buttonList.add(volume = new GuiSlideControl(4, getCenter() - 45, getRowPos(5) + 10, 90, 18, "Volume: ", 0.0f, 30.0f, mod.getSettings().getVolume(), true));
        buttonList.add(new GuiTransButton(5, getCenter() - 45, height - 20, 90, 20, "Save"));
    }

    public int getRowPos(int rowNumber) {
        return height / 4 + (24 * rowNumber - 24) - 16;
    }

    public int getCenter() {
        return width / 2;
    }

    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        int yPosTitle = 10;
        super.drawDefaultBackground();
        GL11.glPushMatrix();
        GL11.glTranslatef(1.0f * -getCenter(), 1.0f * -yPosTitle, 0.0f);
        GL11.glScaled(2.0, 2.0, 2.0);
        drawCenteredString(mc.fontRendererObj, "OofMod V2", getCenter(), yPosTitle, 815000);
        GL11.glPopMatrix();
        GL11.glPushMatrix();
        GL11.glTranslatef(-0.3f * -getCenter(), -0.3f * -yPosTitle + 12.0f, 0.0f);
        GL11.glScaled(0.7, 0.7, 0.7);
        drawCenteredString(mc.fontRendererObj, "By Refraction & powns, ported by chachy", getCenter(), yPosTitle + 12, 815000);
        GL11.glPopMatrix();
        super.drawScreen(mouseX, mouseY, partialTicks);
        renderUI();
        renderWavIcon();
    }

    protected void actionPerformed(GuiButton button) {
        switch (button.id) {
            case 0: {
                ArrayList<File> allSounds = mod.getSettings().getSounds();
                if (allSounds.size() <= 1) {
                    return;
                }
                if (currentIndex - 1 < 0) {
                    currentIndex = allSounds.size() - 1;
                    break;
                }
                --currentIndex;
                break;
            }
            case 1: {
                ArrayList<File> allSounds = mod.getSettings().getSounds();
                if (allSounds.size() <= 1) {
                    return;
                }
                if (currentIndex + 1 > allSounds.size() - 1) {
                    currentIndex = 0;
                    break;
                }
                ++currentIndex;
                break;
            }
            case 2: {
                try {
                    previewSound(mod.getSettings().getSounds().get(currentIndex).getName(), volume.getValueAsFloat() - 30.0f);
                } catch (Exception ignored) {
                }
                break;
            }
            case 3: {
                mod.getSettings().setEnabled(!mod.getSettings().isEnabled());
                button.displayString = (mod.getSettings().isEnabled() ? (EnumChatFormatting.GREEN + "Enabled") : (EnumChatFormatting.RED + "Disabled"));
                break;
            }
            case 5: {
                mod.getSettings().setVolume(volume.getValueAsFloat());
                mod.getSettings().setSelectedSoundName(mod.getSettings().getSounds().get(currentIndex).getName());
                Hyperium.CONFIG.save();
                mc.displayGuiScreen(null);
                break;
            }
        }
    }

    protected void renderUI() {
        String version = "2.2";
        float f2 = 0.4862745f;
        float f3 = 0.6392157f;
        float f4 = 0.7137255f;
        if (width >= 640 && height >= 350) {
            GL11.glPushMatrix();
            GL11.glScaled(1.0, 1.0, 0.0);
            drawString(mc.fontRendererObj, "Version: " + Minecraft.getMinecraft().getVersion() + " - " + version, 2, 2, 8168374);
            GL11.glPopMatrix();
            GL11.glPushMatrix();
            GL11.glEnable(3042);
            GL11.glDisable(3553);
            GL11.glBlendFunc(770, 771);
            GL11.glColor4f(f2, f3, f4, 0.3f);
            Tessellator tes = Tessellator.getInstance();
            WorldRenderer worldrenderer = tes.getWorldRenderer();
            worldrenderer.begin(4, DefaultVertexFormats.POSITION);
            worldrenderer.pos(0.0, 0.0, 0.0).endVertex();
            worldrenderer.pos(0.0, 140.0, 0.0).endVertex();
            worldrenderer.pos(140.0, 0.0, 0.0).endVertex();
            tes.draw();
            GL11.glEnable(3553);
            GL11.glDisable(3042);
            GL11.glPopMatrix();
            GL11.glPushMatrix();
            GL11.glEnable(3042);
            GL11.glDisable(3553);
            GL11.glBlendFunc(770, 771);
            GL11.glColor4f(f2, f3, f4, 0.3f);
            Tessellator tes2 = Tessellator.getInstance();
            WorldRenderer worldrenderer2 = tes2.getWorldRenderer();
            worldrenderer2.begin(4, DefaultVertexFormats.POSITION);
            worldrenderer2.pos(width, height, 0.0).endVertex();
            worldrenderer2.pos(width, height - 140, 0.0).endVertex();
            worldrenderer2.pos(width - 140, height, 0.0).endVertex();
            tes.draw();
            GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
            GL11.glEnable(3553);
            GL11.glDisable(3042);
            GL11.glPopMatrix();
        }
    }

    private void renderWavIcon() {
        GlStateManager.pushMatrix();
        GlStateManager.enableBlend();
        GlStateManager.enableAlpha();
        GlStateManager.color(1.0f, 1.0f, 1.0f);
        ResourceLocation location = new ResourceLocation("wav-icon.png");
        mc.getRenderManager().renderEngine.bindTexture(location);
        Gui.drawModalRectWithCustomSizedTexture(getCenter() - 20, getRowPos(1) + 10, 0.0f, 0.0f, 40, 51, 40.0f, 51.0f);
        GlStateManager.popMatrix();
        GL11.glPushMatrix();
        GL11.glTranslatef(-0.19999999f * -getCenter(), -0.19999999f * -(getRowPos(4) - 6), 0.0f);
        GL11.glScaled(0.8, 0.8, 0.8);
        drawCenteredString(mc.fontRendererObj, mod.getSettings().getSounds().get(currentIndex).getName(), getCenter(), getRowPos(4) - 6, -1);
        GL11.glPopMatrix();
    }

    private void previewSound(String fileName, float volume) throws Exception {
        File soundFile = new File("config/oofmod/sounds/" + fileName);
        AudioInputStream audioIn = AudioSystem.getAudioInputStream(soundFile.toURI().toURL());
        Clip clip = AudioSystem.getClip();
        clip.open(audioIn);
        FloatControl gainControl = (FloatControl) clip.getControl(FloatControl.Type.MASTER_GAIN);
        gainControl.setValue(volume);
        clip.start();
    }

    private int getColorWithAlpha(int rgb, int a) {
        int r = rgb >> 16 & 0xFF;
        int g = rgb >> 8 & 0xFF;
        int b = rgb & 0xFF;
        return a << 24 | r << 16 | g << 8 | b;
    }

    public void onGuiClosed() {
        Hyperium.CONFIG.save();
    }
}